pcap/pcap.go

```diff
@@ -0,0 +1,311 @@
+	"golang.org/x/crypto/ssh"
+
+	. "github.com/cloudfoundry/bosh-cli/v7/cmd/opts"
+	boshdir "github.com/cloudfoundry/bosh-cli/v7/director"
+	boshui "github.com/cloudfoundry/bosh-cli/v7/ui"
+)
+
+var (
+	ErrValidationFailed = fmt.Errorf("validation failed")
+	ErrIllegalCharacter = fmt.Errorf("illegal character: %w", ErrValidationFailed)
+	openHandleError     = fmt.Errorf("unable to open pcap handle")
+)
+
+// You only need **one** of these per package!
+//go:generate go run github.com/maxbrunsfeld/counterfeiter/v6 -generate
+
+//counterfeiter:generate . PcapRunner
+type PcapRunner interface {
+	Run(boshdir.SSHResult, string, string, PcapOpts, ssh.Signer) error
+}
+
+func NewPcapRunner(ui boshui.UI) PcapRunner {
+	return PcapRunnerImpl{ui: ui}
+}
+
+type PcapRunnerImpl struct {
+	ui boshui.UI
+}
+
+func (p PcapRunnerImpl) Run(result boshdir.SSHResult, username string, argv string, opts PcapOpts, privateKey ssh.Signer) error {
+	var packetCs []<-chan gopacket.Packet
+
+	done := make(chan struct{})
+
+	wg := &sync.WaitGroup{}
+
+	ctx, cancel := context.WithCancelCause(context.Background())
+
+	runningCaptures := 0
+
+	for _, host := range result.Hosts {
+
+		p.ui.BeginLinef("Start capture on %s/%s\n", host.Job, host.IndexOrID)
+
+		packets, err := captureSSH(argv, opts.Filter, username, host, privateKey, opts.StopTimeout, wg, done, p.ui, ctx, cancel)
+		if err != nil {
+			// c.ui.ErrorLinef writes error message to stdout/stderr but does not stop the workflow
+			p.ui.ErrorLinef("Capture cannot be started on the instance: %s/%s due to error. \nContinue on other instances", host.Job, host.IndexOrID)
+
+			continue
+		}
+
+		runningCaptures++
+
+		packetCs = append(packetCs, packets)
+	}
+
+	if runningCaptures == 0 {
+		return fmt.Errorf("starting of all pcap captures failed")
+	}
+
+	err := writePacketsToFile(opts.SnapLength, opts.Output, packetCs, p.ui)
+	if err != nil {
+		return fmt.Errorf("write to output file failed: %w", err)
+	}
+
+	signals := make(chan os.Signal, 1)
+	signal.Notify(signals, syscall.SIGTERM, syscall.SIGINT, syscall.SIGQUIT)
+
+	select {
+	case <-signals:
+		close(done)
+	case <-ctx.Done():
+		// ctx canceled as cmd exited or an error occurred
+	}
+
+	p.ui.BeginLinef("Wait for session has been finished\n")
+
+	wg.Wait()
+
+	p.ui.EndLinef("\nCapture finished")
+
+	return nil
+}
+
+func writePacketsToFile(snapLength uint32, outputFile string, packetCs []<-chan gopacket.Packet, ui boshui.UI) error {
+	// setup output pcap-file
+	packetFile, err := os.Create(outputFile)
+	if err != nil {
+		return err
+	}
+	packetWriter := pcapgo.NewWriter(packetFile)
+	err = packetWriter.WriteFileHeader(snapLength, layers.LinkTypeEthernet)
+	if err != nil {
+		return err
+	}
+
+	mergedPackets := mergePackets(packetCs)
+	go func() {
+		for packet := range mergedPackets {
+			err = packetWriter.WritePacket(packet.Metadata().CaptureInfo, packet.Data())
+			if err != nil {
+				ui.ErrorLinef("Writing packet to file failed")
```

Comment:
I'm not familiar enough to know - are errors from the packet writer recoverable? Should we expose the error message here?